src/common/async/bind_handler.h

```diff
@@ -0,0 +1,111 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2018 Red Hat
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#ifndef CEPH_ASYNC_BIND_HANDLER_H
+#define CEPH_ASYNC_BIND_HANDLER_H
+
+#include <tuple>
+#include <boost/asio.hpp>
+
+namespace ceph::async {
+
+/**
+ * A bound completion handler for use with boost::asio.
+ *
+ * A completion handler wrapper that allows a tuple of arguments to be forwarded
+ * to the original Handler. This is intended for use with boost::asio functions
+ * like defer(), dispatch() and post() which expect handlers which are callable
+ * with no arguments.
+ *
+ * The original Handler's associated allocator and executor are maintained.
+ *
+ * @see bind_handler
+ */
+template <typename Handler, typename Tuple>
+struct CompletionHandler {
+  Handler handler;
+  Tuple args;
+
+  CompletionHandler(Handler&& handler, Tuple&& args)
+    : handler(std::move(handler)),
+      args(std::move(args))
+  {}
+
+  void operator()() & {
```

Comment:
Great that you have these bases covered! This enables CompletionHandler objects to be called in a variety of contexts and do the right thing with their handler and args. Yay!

Reply:
thanks for the reminder - i've actually gone back and forth on this one, and could use some advice

both asio and beast have a similar mechanism (`boost::asio::detail::bind_handler()` and `boost::beast::bind_handler()`), but without the `&&` overload. and there's still a missing piece that makes this `&&` overload useful, because executors will always invoke handlers as an lvalue reference

my first intuition was to implement this `operator() &` overload with `std::move(args)` as a workaround, but that would be destructive to the arguments, and you wouldn't be able to use this class as a general _reusable_ function object

to actually make use of this `&&` overload with executors, you would need an extra wrapper that converts the `operator() &` into `operator() &&`, which would look something like `forward_handler(bind_handler(h, args...))`

so `forward_handler()` would be a requirement for async operations with a move-only type in their Signature, i.e. `void(error_code, unique_ptr<T>)` - and `async::Completion` would need to do that also because it's generic over all Signatures

that would allow `SharedMutex::async_lock_shared()` to use a `Signature = void(error_code, std::shared_lock)` for a more natural interface:
```diff
-    pause_mutex.async_lock_shared(yield[ec]);
+    auto lock = pause_mutex.async_lock_shared(yield[ec]);
     if (ec) {
       return;
     }
-    std::shared_lock lock{pause_mutex, std::adopt_lock};
```

the question, though, is whether this extra layer of wrapping with `forward_handler()` is actually worth it, or if this `operator() &` overload really should just `std::move(args)`? i'm leaning towards the latter now, because i don't think this `CompletionHandler` is all that useful outside of the fact that it preserves the associated allocator/executor - `std::bind()` is better in every other way (supports partial binding, placeholders, etc)

Reply:
thanks, that helps. i'm happy with the composition of forward_handler and bind_handler here. both can be optimized out anyway (see https://godbolt.org/g/oeokp8), so it doesn't cost us anything